### PR TITLE
Merge branch `feature/abci++ppp` into `main`

### LINKED ISSUE #9300: Merge the `feature/abci++ppp` branch into `main` once all relevant work from #9053 is done
As per title

### PR DESCRIPTION
**Pull request by sergio-mena**
Closes #9300

This PR is a merging branch `feature/abci++ppp` into `main`

After this PR is in, branch `feature/abci++ppp` will be removed (as its history is now merged into the main branch).

The contents of this PR represent the whole set of changes backported from `v0.35.x`/`v0.36.x` to deliver `PrepareProposal`/`ProcessProposal` in the next release.

(the CHANGELOG_PENDING.md update is coming in a PR to be created shortly)

---

#### PR checklist

- [x] Tests written/updated, or no tests needed
- [ ] `CHANGELOG_PENDING.md` updated, or no changelog entry needed
- [x] Updated relevant documentation (`docs/`) and code comments, or no
      documentation updates needed

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -51,7 +51,6 @@ proto/spec/**/*.pb.go
 *.aux
 *.bbl
 *.blg
-*.log
```
Comment (tychoish) [inline .gitignore:54]: I don't think that we want to delete this

File: CHANGELOG_PENDING.md [hunk 2]
```diff
@@ -17,6 +17,12 @@ Friendly reminder, we have a [bug bounty program](https://hackerone.com/tendermi
   - [abci/counter] \#6684 Delete counter example app
   - [txResults] \#9175 Remove `gas_used` & `gas_wanted` from being merkelized in the lastresulthash in the header
   - [abci] \#5783 Make length delimiter encoding consistent (`uint64`) between ABCI and P2P wire-level protocols
+  - [abci] \#9145 Removes unused Response/Request `SetOption` from ABCI (@samricotta)
+  - [abci] \#8656 Added cli command for `PrepareProposal`. (@jmalicevic)
+  - [abci] \#8901 Added cli command for `ProcessProposal`. (@hvanz)
```
Comment (tychoish) [inline CHANGELOG_PENDING.md:22]: can we combine these lines

File: abci/example/kvstore/kvstore.go [hunk 3]
```diff
@@ -170,3 +181,18 @@ func (app *Application) Query(reqQuery types.RequestQuery) (resQuery types.Respo
...
 	return resQuery
 }
+
+func (app *Application) BeginBlock(req types.RequestBeginBlock) types.ResponseBeginBlock {
+	app.txToRemove = map[string]struct{}{}
+	return types.ResponseBeginBlock{}
+}
+
+func (app *Application) ProcessProposal(
+	req types.RequestProcessProposal) types.ResponseProcessProposal {
```
Comment (tychoish) [inline abci/example/kvstore/kvstore.go:191]: I think splitting this makes it harder to read, and we haven't typically enforced line lengths.

**Review by tychoish — COMMENTED** [review]
this all seems good, just some minor quibbles. 

I would have expected that the SetOption removal pieces would have been able to happen in an orthogonal PR, but it seems like this is fine too.

File: .gitignore (hunk 1 above)
Comment (sergio-mena) [reply .gitignore:54]: We do, as it already exists in that file

File: CHANGELOG_PENDING.md (hunk 2 above)
Comment (sergio-mena) [reply CHANGELOG_PENDING.md:22]: I am preparing a CHANGELOG PR spanning all `PrepareProposal`/`ProcessProposal` changes that will deal with this

File: abci/example/kvstore/kvstore.go (hunk 3 above)
Comment (sergio-mena) [reply abci/example/kvstore/kvstore.go:191]: As this is a merge commit, I prefer not to tackle these minor issues now, but in a subsequent PR.
Recall this is already reviewed code (on the feature branch), and someone decided this was good to go.

**Comment by sergio-mena** [comment]
> I would have expected that the SetOption removal pieces would have been able to happen in an orthogonal PR, but it seems like this is fine too.

Yeah we hesitated whether to base the PR on `main` or on `feature/abci++ppp`. In the end we went for the latter, as other changes in ABCI's types.proto where going on, and wanted to avoid conflicts at the moment of merging the branch into `main` (i.e., now)

**Review by thanethomson — APPROVED** [review]

**Pull request merged**